content/advent-2019/directional-channels.md

```diff
@@ -0,0 +1,193 @@
+
+```go
+func takesReadonly(c <-chan string){
+    // c is now receive-only inside the function and anywhere else it might go from here
+}
+
+func returnsReadOnly() <-chan string{
+    c := make(chan string)
+    go func(){
+        // some concurrent work with c
+    }()
+    return c
+}
+readOnly := returnsReadOnly()
+
+```
+
+This is a pretty nifty trick, and works a bit differently to conversions in the rest of the language, but, most crucially, the change in direction is reflected in the *type*, which means these restrictions can be enforced at *compile time*.
+
+```go
+go func() {
+    biDirectional <- "hello" // no problem
+    close(biDirectional)     // totally fine
+}()
+go func() {
+    readOnly <- "hello" //"invalid operation ... (send to receive-only type <-chan string)"
+    close(readOnly)     //"invalid operation: ... (cannot close receive-only channel)"
+}()
+
+fmt.Printf("%T\n", readOnly) // "<-chan string" (different type)
+fmt.Println(<-readOnly)      // "hello" (same underlying channel!)
+```
+[playground](https://play.golang.org/p/y1xe8R9wQHK)
+
+This is useful not only to control who can write to or close your channel, but also in terms of descriptiveness and Intentionality. One of the nice things about strongly-typed languages like Go is that they can be tremendously descriptive just through their API. Take the following function as an example:
+
+```go
+func SliceIterChan(s []int) <-chan int {}
+```
+
+Even without the documentation or implementation, this code unambiguously states that it returns a channel that the consumer is supposed to read from, either forever, or until it's closed (which documentation can help clarify). This lends itself very well to a **for-range** over the provided channel.
+
+```go
+for i := range SliceIterChan(someSlice) {
+    fmt.Printf("got %d from channel\n", i)
+}
+fmt.Println("channel closed!")
+```
+
+Diving into the implementation, the function creates a bidirectional channel for its own use, and then all it needs to do to ensure that it has full control over writing to and closing the channel is to return it, whereupon it will be converted into a read-only channel automatically.
+
+```go
+// SliceIterChan returns each element of a slice on a channel for concurrent
+// consumption, closing the channel on completion
+func SliceIterChan(s []int) <-chan int {
+	outChan := make(chan int)
+	go func() {
+		for i := range s {
+			outChan <- s[i]
+		}
+		close(outChan)
+	}()
+	return outChan
+}
+```
+[playground](https://play.golang.org/p/nGMksaNgxAg)
+
+This is a very powerful technique for asserting control over a channel at an API boundary, and one that comes with no cost or need for explicit conversion, beyond simply specifying the channel direction in a declaration. This is so useful, you should use probably use it wherever you return a channel for reading from, unless there's a very good reason not to.
```

Comment:
"This is useful anywhere a channel is returned for the caller to read.  Adding the channel direction makes usage clear and eliminates the possibility of closing from the wrong end."